core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -204,7 +202,117 @@ Map<TopicIdPartition, FetchRequest.PartitionData> acquirablePartitions() {
         return topicPartitionData;
     }
 
-    private void releasePartitionLocks(String groupId, Set<TopicIdPartition> topicIdPartitions) {
+    // In case, fetch offset metadata doesn't exist for any topic partition in the list of topic partitions, we do a
+    // replicaManager.readFromLog to populate the offset metadata.
+    private void maybeUpdateFetchOffsetMetadataForTopicPartitions(Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData) {
+        Map<TopicIdPartition, FetchPartitionOffsetData> replicaManagerReadResponseData = null;
+        for (Map.Entry<TopicIdPartition, FetchRequest.PartitionData> entry : topicPartitionData.entrySet()) {
+            TopicIdPartition topicIdPartition = entry.getKey();
+            SharePartition sharePartition = sharePartitionManager.sharePartition(shareFetchData.groupId(), topicIdPartition);
+            if (sharePartition == null) {
+                log.error("Encountered null share partition for groupId={}, topicIdPartition={}. Skipping it.", shareFetchData.groupId(), topicIdPartition);
+                continue;
+            }
+            if (sharePartition.latestFetchOffsetMetadata() == null) {
```

Comment:
Perhaps it's clearer to make sharePartition.latestFetchOffsetMetadata() an Optional?